# Combining data sets 2
This code includes an example to combine data sets when there are inconsistencies in key variables.

We are using following two data sets in this code. 
* **World band project data:** This data set includes World Bank's leading projects since 1947. You can download this data from [here](https://datacatalog.worldbank.org/dataset/world-bank-projects-operations)  
* **World bank indicator data:** This data set includes World Bank's population. You can download this data from [here](https://data.worldbank.org/indicator/SP.POP.TOTL)

The detailed code for extracting these data sets are given in wrangling_csv notebook in this repository

In [1]:
# first import pandas
import pandas as pd

In [5]:
#Read data sets: Project data
Projects= pd.read_csv("Projects.csv", dtype= str)
Threshold= int(Projects.shape[0]/2)
Projects=Projects.dropna(axis=1, thresh=Threshold)
Projects.head()

id               regionname  \
0  P169599  Europe and Central Asia   
1  P173950                   Africa   
2  P170934                   Africa   
3  P171006    East Asia and Pacific   
4  P171681    East Asia and Pacific   

                                         countryname prodline  \
0                                    Georgia;Georgia       RE   
1      Republic of Madagascar;Republic of Madagascar       PE   
2                          Burkina Faso;Burkina Faso       PE   
3  Socialist Republic of Vietnam;Socialist Republ...       PE   
4                                      Tuvalu;Tuvalu       PE   

                   lendinginstr lendinginstrtype envassesmentcategorycode  \
0  Investment Project Financing               IN                      NaN   
1  Investment Project Financing               IN                      NaN   
2    Development Policy Lending               AD                      NaN   
3    Development Policy Lending               AD                      NaN   
4  Investment Project Financing               IN                      NaN   

  supplementprojectflg productlinetype projectstatusdisplay  ... borrower  \
0                    N               L               Active  ...      NaN   
1                    Y               L               Active  ...      NaN   
2                    N               L               Active  ...      NaN   
3                    N               L               Active  ...      NaN   
4                    N               L               Active  ...      NaN   

  impagency                                                url sector1  \
0       NaN  http://projects.worldbank.org/P169599/null?lan...    !$!0   
1       NaN  http://projects.worldbank.org/P173950/null?lan...    !$!0   
2       NaN  http://projects.worldbank.org/P170934/null?lan...    !$!0   
3       NaN  http://projects.worldbank.org/P171006/null?lan...    !$!0   
4       NaN  http://projects.worldbank.org/P171681/null?lan...    !$!0   

  sector2 sector mjsector theme1 theme2 goal  
0     NaN    NaN      NaN   !$!0    NaN  NaN  
1     NaN    NaN      NaN   !$!0    NaN  NaN  
2     NaN    NaN      NaN   !$!0    NaN  NaN  
3     NaN    NaN      NaN   !$!0    NaN  NaN  
4     NaN    NaN      NaN   !$!0    NaN  NaN  

[5 rows x 30 columns]

In [9]:
Population= pd.read_csv("Population.csv",skiprows=4)
Population= Population.drop(['2019', 'Unnamed: 64'], axis=1)
Population.head()

Country Name Country Code     Indicator Name Indicator Code       1960  \
0        Aruba          ABW  Population, total    SP.POP.TOTL    54211.0   
1  Afghanistan          AFG  Population, total    SP.POP.TOTL  8996973.0   
2       Angola          AGO  Population, total    SP.POP.TOTL  5454933.0   
3      Albania          ALB  Population, total    SP.POP.TOTL  1608800.0   
4      Andorra          AND  Population, total    SP.POP.TOTL    13411.0   

        1961       1962       1963       1964       1965  ...        2009  \
0    55438.0    56225.0    56695.0    57032.0    57360.0  ...    101455.0   
1  9169410.0  9351441.0  9543205.0  9744781.0  9956320.0  ...  28394813.0   
2  5531472.0  5608539.0  5679458.0  5735044.0  5770570.0  ...  22514281.0   
3  1659800.0  1711319.0  1762621.0  1814135.0  1864791.0  ...   2927519.0   
4    14375.0    15370.0    16412.0    17469.0    18549.0  ...     84463.0   

         2010        2011        2012        2013        2014        2015  \
0    101669.0    102046.0    102560.0    103159.0    103774.0    104341.0   
1  29185507.0  30117413.0  31161376.0  32269589.0  33370794.0  34413603.0   
2  23356246.0  24220661.0  25107931.0  26015780.0  26941779.0  27884381.0   
3   2913021.0   2905195.0   2900401.0   2895092.0   2889104.0   2880703.0   
4     84449.0     83747.0     82427.0     80774.0     79213.0     78011.0   

         2016        2017        2018  
0    104872.0    105366.0    105845.0  
1  35383128.0  36296400.0  37172386.0  
2  28842484.0  29816748.0  30809762.0  
3   2876101.0   2873457.0   2866376.0  
4     77297.0     77001.0     77006.0  

[5 rows x 63 columns]

In population df there are two colums identifying the country : Country name and Country Code. Lets first clean the duplicate rows if there were any.

In [11]:
#Drop duplicate rows in population data frame
Population[['Country Name', 'Country Code']].drop_duplicates()

Country Name Country Code
0           Aruba          ABW
1     Afghanistan          AFG
2          Angola          AGO
3         Albania          ALB
4         Andorra          AND
..            ...          ...
259        Kosovo          XKX
260   Yemen, Rep.          YEM
261  South Africa          ZAF
262        Zambia          ZMB
263      Zimbabwe          ZWE

[264 rows x 2 columns]

As seen above there are 264 countries in population data. In projects data the countryname identifies the country information. Lets take a look at the which countries are in the data set. 

In [12]:
Projects['countryname'].unique()

array(['Georgia;Georgia', 'Republic of Madagascar;Republic of Madagascar',
       'Burkina Faso;Burkina Faso',
       'Socialist Republic of Vietnam;Socialist Republic of Vietnam',
       'Tuvalu;Tuvalu',
       'Federated States of Micronesia;Federated States of Micronesia',
       'Kingdom of Tonga;Kingdom of Tonga',
       'Republic of The Gambia;Republic of The Gambia',
       'St. Lucia;St. Lucia',
       'Islamic Republic of Pakistan;Islamic Republic of Pakistan',
       'Republic of Benin;Republic of Benin', 'Nepal;Nepal',
       'Republic of Maldives;Republic of Maldives',
       'Republic of Guinea;Republic of Guinea',
       'Kingdom of Cambodia;Kingdom of Cambodia',
       'Republic of Djibouti;Republic of Djibouti',
       'Republic of Zambia;Republic of Zambia',
       'Republic of the Union of Myanmar;Republic of the Union of Myanmar',
       'Republic of Haiti;Republic of Haiti',
       'Commonwealth of Dominica;Commonwealth of Dominica',
       'Africa;Africa', 'Grenada

As seen above the country name column is a little messy. Each country name was written twice and seperated by ';'. First we need to get rid of the duplicate ones and ; 

In [14]:
#this line creates a new variable named offical country name
Projects['Official Country Name'] = Projects['countryname'].str.split(';').str.get(0)
Projects.head(5)

id               regionname  \
0  P169599  Europe and Central Asia   
1  P173950                   Africa   
2  P170934                   Africa   
3  P171006    East Asia and Pacific   
4  P171681    East Asia and Pacific   

                                         countryname prodline  \
0                                    Georgia;Georgia       RE   
1      Republic of Madagascar;Republic of Madagascar       PE   
2                          Burkina Faso;Burkina Faso       PE   
3  Socialist Republic of Vietnam;Socialist Republ...       PE   
4                                      Tuvalu;Tuvalu       PE   

                   lendinginstr lendinginstrtype envassesmentcategorycode  \
0  Investment Project Financing               IN                      NaN   
1  Investment Project Financing               IN                      NaN   
2    Development Policy Lending               AD                      NaN   
3    Development Policy Lending               AD                      NaN   
4  Investment Project Financing               IN                      NaN   

  supplementprojectflg productlinetype projectstatusdisplay  ... impagency  \
0                    N               L               Active  ...       NaN   
1                    Y               L               Active  ...       NaN   
2                    N               L               Active  ...       NaN   
3                    N               L               Active  ...       NaN   
4                    N               L               Active  ...       NaN   

                                                 url sector1 sector2 sector  \
0  http://projects.worldbank.org/P169599/null?lan...    !$!0     NaN    NaN   
1  http://projects.worldbank.org/P173950/null?lan...    !$!0     NaN    NaN   
2  http://projects.worldbank.org/P170934/null?lan...    !$!0     NaN    NaN   
3  http://projects.worldbank.org/P171006/null?lan...    !$!0     NaN    NaN   
4  http://projects.worldbank.org/P171681/null?lan...    !$!0     NaN    NaN   

  mjsector theme1 theme2 goal          Official Country Name  
0      NaN   !$!0    NaN  NaN                        Georgia  
1      NaN   !$!0    NaN  NaN         Republic of Madagascar  
2      NaN   !$!0    NaN  NaN                   Burkina Faso  
3      NaN   !$!0    NaN  NaN  Socialist Republic of Vietnam  
4      NaN   !$!0    NaN  NaN                         Tuvalu  

[5 rows x 31 columns]

In [19]:
Projects.shape[0]

19361

In [24]:
pd_merged=pd.merge(Population, Projects, left_on='Country Name', right_on='Official Country Name', how='inner')

In [25]:
pd_merged.shape[0]

2210

Even though the merge method did not give us an error, when we check the number of rows we see that there is a problem. There are 19361 rows in projects data but there are only 2210 rows in merged one. There is probably a matching error in country names here.  

While working with country names we can use pycountry library. This library has country names, ISO abbreviations, and official country names. While you might not be able to clean all of the data with the help of this Python library, it will probably help.  

We can use two methods in this library countries.get and countries.lookup. Lets take a look at both


In [33]:
# First we need to install and import the pycountry library
!pip install pycountry
from pycountry import countries

In [34]:
#we can take a look at how this library works
countries.get(name='Italy')

Country(alpha_2='IT', alpha_3='ITA', name='Italy', numeric='380', official_name='Italian Republic')

As seen above there are two abbreviations, country name, numreic code and offical name for each country. We can use this library to create a country code variable.  
We will create a dictionary for country abbreviations for the country names in project data frame

In [35]:
countries.lookup('Italian Republic')

Country(alpha_2='IT', alpha_3='ITA', name='Italy', numeric='380', official_name='Italian Republic')

In [31]:
# initialize an empty list and dictionary
from collections import defaultdict
country_not_found = [] # stores countries not found in the pycountry library
project_country_abbrev_dict = defaultdict(str) # set up an empty dictionary of string values


for country in Projects['Official Country Name'].drop_duplicates().sort_values():
    try: 
        project_country_abbrev_dict[country] = countries.lookup(country).alpha_3
    except:
        print(country, ' not found')
        country_not_found.append(country)

Africa  not found
Andean Countries  not found
Aral Sea  not found
Asia  not found
Caribbean  not found
Caucasus  not found
Central Africa  not found
Central America  not found
Central Asia  not found
Co-operative Republic of Guyana  not found
Commonwealth of Australia  not found
Democratic Republic of Sao Tome and Prin  not found
Democratic Republic of the Congo  not found
Democratic Socialist Republic of Sri Lan  not found
EU Accession Countries  not found
East Asia and Pacific  not found
Eastern Africa  not found
Europe and Central Asia  not found
Islamic  Republic of Afghanistan  not found
Latin America  not found
Macedonia  not found
Mekong  not found
Mercosur  not found
Middle East and North Africa  not found
Multi-Regional  not found
Organization of Eastern Caribbean States  not found
Oriental Republic of Uruguay  not found
Pacific Islands  not found
Red Sea and Gulf of Aden  not found
Republic of Congo  not found
Republic of Cote d'Ivoire  not found
Republic of Korea  not found


We have some countries that are not found in the library. We can check whether these names matches with the names in population data

In [36]:
indicator_countries = Population[['Country Name', 'Country Code']].drop_duplicates().sort_values(by='Country Name')

for country in country_not_found:
    if country in indicator_countries['Country Name'].tolist():
        print(country)

South Asia
St. Kitts and Nevis
St. Lucia
St. Vincent and the Grenadines
West Bank and Gaza
World


In [42]:
# we can manually create adictionay for the not found ones

country_not_found_mapping = {'Co-operative Republic of Guyana': 'GUY',
             'Commonwealth of Australia':'AUS',
             'Democratic Republic of Sao Tome and Prin':'STP',
             'Democratic Republic of the Congo':'COD',
             'Democratic Socialist Republic of Sri Lan':'LKA',
             'East Asia and Pacific':'EAS',
             'Europe and Central Asia': 'ECS',
             'Islamic  Republic of Afghanistan':'AFG',
             'Latin America':'LCN',
              'Caribbean':'LCN',
             'Macedonia':'MKD',
             'Middle East and North Africa':'MEA',
             'Oriental Republic of Uruguay':'URY',
             'Republic of Congo':'COG',
             "Republic of Cote d'Ivoire":'CIV',
             'Republic of Korea':'KOR',
             'Republic of Niger':'NER',
             'Republic of Kosovo':'XKX',
             'Republic of Rwanda':'RWA',
              'Republic of The Gambia':'GMB',
              'Republic of Togo':'TGO',
              'Republic of the Union of Myanmar':'MMR',
              'Republica Bolivariana de Venezuela':'VEN',
              'Sint Maarten':'SXM',
              "Socialist People's Libyan Arab Jamahiriy":'LBY',
              'Socialist Republic of Vietnam':'VNM',
              'Somali Democratic Republic':'SOM',
              'South Asia':'SAS',
              'St. Kitts and Nevis':'KNA',
              'St. Lucia':'LCA',
              'St. Vincent and the Grenadines':'VCT',
              'State of Eritrea':'ERI',
              'The Independent State of Papua New Guine':'PNG',
              'West Bank and Gaza':'PSE',
              'World':'WLD'}

In [43]:
#lets add this dictionay to our abbrev_dict
project_country_abbrev_dict.update(country_not_found_mapping)

In [44]:
#This code creates a column name Country code in projects data
Projects['Country Code'] = Projects['Official Country Name'].apply(lambda x: project_country_abbrev_dict[x])

In [45]:
Projects.head()

id               regionname  \
0  P169599  Europe and Central Asia   
1  P173950                   Africa   
2  P170934                   Africa   
3  P171006    East Asia and Pacific   
4  P171681    East Asia and Pacific   

                                         countryname prodline  \
0                                    Georgia;Georgia       RE   
1      Republic of Madagascar;Republic of Madagascar       PE   
2                          Burkina Faso;Burkina Faso       PE   
3  Socialist Republic of Vietnam;Socialist Republ...       PE   
4                                      Tuvalu;Tuvalu       PE   

                   lendinginstr lendinginstrtype envassesmentcategorycode  \
0  Investment Project Financing               IN                      NaN   
1  Investment Project Financing               IN                      NaN   
2    Development Policy Lending               AD                      NaN   
3    Development Policy Lending               AD                      NaN   
4  Investment Project Financing               IN                      NaN   

  supplementprojectflg productlinetype projectstatusdisplay  ...  \
0                    N               L               Active  ...   
1                    Y               L               Active  ...   
2                    N               L               Active  ...   
3                    N               L               Active  ...   
4                    N               L               Active  ...   

                                                 url sector1 sector2 sector  \
0  http://projects.worldbank.org/P169599/null?lan...    !$!0     NaN    NaN   
1  http://projects.worldbank.org/P173950/null?lan...    !$!0     NaN    NaN   
2  http://projects.worldbank.org/P170934/null?lan...    !$!0     NaN    NaN   
3  http://projects.worldbank.org/P171006/null?lan...    !$!0     NaN    NaN   
4  http://projects.worldbank.org/P171681/null?lan...    !$!0     NaN    NaN   

  mjsector theme1 theme2 goal          Official Country Name Country Code  
0      NaN   !$!0    NaN  NaN                        Georgia          GEO  
1      NaN   !$!0    NaN  NaN         Republic of Madagascar          MDG  
2      NaN   !$!0    NaN  NaN                   Burkina Faso          BFA  
3      NaN   !$!0    NaN  NaN  Socialist Republic of Vietnam          VNM  
4      NaN   !$!0    NaN  NaN                         Tuvalu          TUV  

[5 rows x 32 columns]

In [47]:
#we can check whether there are rows who don't have country codes
Projects[Projects['Country Code'] == '']

id               regionname  \
24     P171827                   Africa   
30     P161706                   Africa   
41     P173702                   Africa   
85     P169594                   Africa   
187    P169265                   Africa   
...        ...                      ...   
19076  P009136  Europe and Central Asia   
19247  P000620                   Africa   
19288  P009135  Europe and Central Asia   
19310  P009134  Europe and Central Asia   
19341  P009133  Europe and Central Asia   

                                             countryname prodline  \
24                                         Africa;Africa       PE   
30                                         Africa;Africa       PE   
41                                         Africa;Africa       PE   
85                         Western Africa;Western Africa       PE   
187                        Western Africa;Western Africa       PE   
...                                                  ...      ...   
19076  Socialist Federal Republic of Yugoslavia;Socia...       PE   
19247                                      Africa;Africa       PE   
19288  Socialist Federal Republic of Yugoslavia;Socia...       PE   
19310  Socialist Federal Republic of Yugoslavia;Socia...       PE   
19341  Socialist Federal Republic of Yugoslavia;Socia...       PE   

                       lendinginstr lendinginstrtype envassesmentcategorycode  \
24     Investment Project Financing               IN                      NaN   
30     Investment Project Financing               IN                        B   
41     Investment Project Financing               IN                      NaN   
85     Investment Project Financing               IN                      NaN   
187    Investment Project Financing               IN                        B   
...                             ...              ...                      ...   
19076      Specific Investment Loan               IN                      NaN   
19247      Specific Investment Loan               IN                        C   
19288    Structural Adjustment Loan               AD                      NaN   
19310    Structural Adjustment Loan               AD                      NaN   
19341      Specific Investment Loan               IN                      NaN   

      supplementprojectflg productlinetype projectstatusdisplay  ...  \
24                       Y               L               Active  ...   
30                       N               L               Active  ...   
41                       N               L               Active  ...   
85                       N               L               Active  ...   
187                      N               L               Active  ...   
...                    ...             ...                  ...  ...   
19076                    N               L               Closed  ...   
19247                    N               L               Closed  ...   
19288                    N               L               Closed  ...   
19310                    N               L               Closed  ...   
19341                    N               L               Closed  ...   

                                                     url  \
24     http://projects.worldbank.org/P171827/null?lan...   
30     http://projects.worldbank.org/P161706/null?lan...   
41     http://projects.worldbank.org/P173702/null?lan...   
85     http://projects.worldbank.org/P169594/null?lan...   
187    http://projects.worldbank.org/P169265/null?lan...   
...                                                  ...   
19076  http://projects.worldbank.org/P009136/electric...   
19247  http://projects.worldbank.org/P000620/railways...   
19288  http://projects.worldbank.org/P009135/power-mi...   
19310  http://projects.worldbank.org/P009134/power-mi...   
19341  http://projects.worldbank.org/P009133/agricult...   

                                               sector1  \
24                                               